File: tests/architecture-audit-checker/javascript/src/parser.js
/**
 * Markdown parser utilities for the architecture-audit checker harness
 * (JavaScript track).
 *
 * Provides three exports that mirror the TypeScript track:
 *
 *   - {@link extractTables}     parse all GFM tables in a Markdown document
 *                               into header/row arrays.
 *   - {@link extractSection}    extract the body of a heading-bounded section
 *                               by exact heading text.
 *   - {@link extractCodeBlocks} return all fenced code blocks, optionally
 *                               filtered by info-string language.
 *
 * Backed by `unified` + `remark-parse` + `remark-gfm` so GFM tables, fenced
 * code blocks, and standard Markdown headings parse identically across the
 * TS and JS tracks.
 */

import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkGfm from "remark-gfm";

/**
 * @typedef {Object} ParsedTable
 * @property {string[]} headers
 *   Trimmed header cell text in document order.
 * @property {string[][]} rows
 *   Each row is an array of trimmed cell strings, length-aligned to headers.
 *   Cells beyond `headers.length` are dropped; missing cells are filled with
 *   the empty string.
 */

/**
 * @typedef {Object} CodeBlock
 * @property {string|null} lang
 *   The fenced code-block info-string language, or null when absent.
 * @property {string} value
 *   Raw text content of the code block (no surrounding fences).
 */

/**
 * Internal helper: build the shared remark processor. We instantiate per-call
 * so concurrent parses cannot interfere with one another's state.
 *
 * @returns {ReturnType<typeof unified>}
 */
function buildProcessor() {
  return unified().use(remarkParse).use(remarkGfm);
}

/**
 * Extract the visible text of a remark AST node and all its descendants. Used
 * to flatten heading and table-cell nodes into plain strings.
 *
 * @param {any} node
 * @returns {string}
 */
function nodeText(node) {
  if (!node) return "";
  if (typeof node.value === "string") return node.value;
  if (Array.isArray(node.children)) {
    return node.children.map(nodeText).join("");
  }
  return "";
}

/**
 * Parse all GFM tables in a Markdown document.
 *
 * @param {string} markdown
 *   Raw Markdown source. An empty string yields an empty array.
 * @returns {ParsedTable[]}
 */
export function extractTables(markdown) {
  if (!markdown) return [];

  const tree = buildProcessor().parse(markdown);
  /** @type {ParsedTable[]} */
  const tables = [];

  for (const node of tree.children ?? []) {
    if (node.type !== "table") continue;

    const tableRows = node.children ?? [];
    if (tableRows.length === 0) continue;

    const headerCells = tableRows[0].children ?? [];
    const headers = headerCells.map((cell) => nodeText(cell).trim());

    const rows = [];
    for (let i = 1; i < tableRows.length; i++) {
      const dataCells = tableRows[i].children ?? [];
      const row = headers.map((_, idx) =>
        idx < dataCells.length ? nodeText(dataCells[idx]).trim() : "",
      );
      rows.push(row);
    }

    tables.push({ headers, rows });
  }

  return tables;
}

/**
 * Extract the body of a section by its exact heading text.
 *
 * The returned slice begins at the first character after the matched heading
 * line and ends just before the next heading whose depth is less than or
 * equal to the matched heading's depth (closing the section), or end of file.
 *
 * @param {string} markdown
 *   Raw Markdown source.
 * @param {string} headingText
 *   The exact heading text to match (case-sensitive, whitespace-trimmed).
 * @returns {string|null}
 *   The section body, or `null` if no matching heading exists.
 */
export function extractSection(markdown, headingText) {
  if (!markdown) return null;

  const tree = buildProcessor().parse(markdown);
  const children = tree.children ?? [];

  let startIdx = -1;
  let startDepth = -1;
  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    if (node.type !== "heading") continue;
    if (nodeText(node).trim() === headingText.trim()) {
      startIdx = i;
      startDepth = node.depth;
      break;
    }
  }
  if (startIdx === -1) return null;

  let endIdx = children.length;
  for (let i = startIdx + 1; i < children.length; i++) {
    const node = children[i];
    if (node.type === "heading" && node.depth <= startDepth) {
      endIdx = i;
      break;
    }
  }

  // Convert the AST-positional bound back to a substring of the source.
  const headingNode = children[startIdx];
  const startOffset = headingNode.position?.end?.offset;
  const endNode = endIdx < children.length ? children[endIdx] : null;
  const endOffset = endNode?.position?.start?.offset ?? markdown.length;

  if (typeof startOffset !== "number") return "";
  return markdown.slice(startOffset, endOffset).replace(/^\n+/, "");
}

/**
 * Return all fenced code blocks in document order.
 *
 * @param {string} markdown
 *   Raw Markdown source.
 * @param {string} [lang]
 *   Optional info-string filter. When provided, only blocks whose `lang`
 *   matches exactly are returned.
 * @returns {CodeBlock[]}
 */
export function extractCodeBlocks(markdown, lang) {
  if (!markdown) return [];

  const tree = buildProcessor().parse(markdown);
  /** @type {CodeBlock[]} */
  const blocks = [];

  for (const node of tree.children ?? []) {
    if (node.type !== "code") continue;
    const blockLang = typeof node.lang === "string" ? node.lang : null;
    if (lang !== undefined && blockLang !== lang) continue;
    blocks.push({ lang: blockLang, value: typeof node.value === "string" ? node.value : "" });
  }

  return blocks;
}

File: tests/architecture-audit-checker/rust/src/parser.rs
//! Thin Markdown parser wrappers built on top of `pulldown-cmark`.
//!
//! The 17 Correctness Properties only need three coarse extractors:
//!
//! 1. [`extract_tables`] — every GFM table in document order, returned as a
//!    list of headers + rows of plain-text cell content.
//! 2. [`extract_section`] — the body text of a section identified by its
//!    heading text, terminated by the next heading at the same level or a
//!    higher (smaller-numbered) level.
//! 3. [`extract_code_blocks`] — fenced code-block bodies, optionally filtered
//!    by the language tag (e.g. `mermaid`, `json`).
//!
//! GFM tables are enabled via [`Options::ENABLE_TABLES`] so that the
//! deliverable Markdown round-trips through this parser without losing
//! structure.

use pulldown_cmark::{CodeBlockKind, Event, HeadingLevel, Options, Parser, Tag, TagEnd};

/// One Markdown table, flattened to plain-text cell strings.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Table {
    pub headers: Vec<String>,
    pub rows: Vec<Vec<String>>,
}

fn options() -> Options {
    let mut opts = Options::empty();
    opts.insert(Options::ENABLE_TABLES);
    opts
}

/// Extract every GFM table from `md` in document order.
pub fn extract_tables(md: &str) -> Vec<Table> {
    let parser = Parser::new_ext(md, options());

    let mut tables: Vec<Table> = Vec::new();
    let mut current: Option<Table> = None;
    let mut in_head = false;
    let mut in_row = false;
    let mut in_cell = false;
    let mut current_row: Vec<String> = Vec::new();
    let mut cell_buf = String::new();

    for event in parser {
        match event {
            Event::Start(Tag::Table(_)) => {
                current = Some(Table {
                    headers: Vec::new(),
                    rows: Vec::new(),
                });
            }
            Event::End(TagEnd::Table) => {
                if let Some(table) = current.take() {
                    tables.push(table);
                }
            }
            Event::Start(Tag::TableHead) => {
                in_head = true;
                current_row = Vec::new();
            }
            Event::End(TagEnd::TableHead) => {
                if let Some(table) = current.as_mut() {
                    table.headers = std::mem::take(&mut current_row);
                }
                in_head = false;
            }
            Event::Start(Tag::TableRow) => {
                in_row = true;
                current_row = Vec::new();
            }
            Event::End(TagEnd::TableRow) => {
                if let Some(table) = current.as_mut() {
                    table.rows.push(std::mem::take(&mut current_row));
                }
                in_row = false;
            }
            Event::Start(Tag::TableCell) => {
                in_cell = true;
                cell_buf.clear();
            }
            Event::End(TagEnd::TableCell) => {
                if in_head || in_row {
                    current_row.push(cell_buf.trim().to_string());
                }
                cell_buf.clear();
                in_cell = false;
            }
            Event::Text(text) | Event::Code(text) | Event::Html(text) => {
                if in_cell {
                    cell_buf.push_str(&text);
                }
            }
            Event::SoftBreak | Event::HardBreak => {
                if in_cell {
                    cell_buf.push(' ');
                }
            }
            _ => {}
        }
    }

    tables
}

/// Extract the body text of the first section whose heading text exactly
/// equals `heading`. The section ends at the next heading whose level is the
/// same as or higher than the matched heading (i.e. `## Foo` is terminated by
/// the next `## ...` or any `# ...`).
///
/// Returns the empty string if no matching heading is found.
pub fn extract_section(md: &str, heading: &str) -> String {
    let parser = Parser::new_ext(md, options()).into_offset_iter();
    let target = heading.trim();

    let mut match_level: Option<HeadingLevel> = None;
    let mut heading_open: Option<HeadingLevel> = None;
    let mut heading_text = String::new();
    let mut start_byte: Option<usize> = None;
    let mut end_byte: Option<usize> = None;

    for (event, range) in parser {
        match event {
            Event::Start(Tag::Heading { level, .. }) => {
                heading_open = Some(level);
                heading_text.clear();
            }
            Event::End(TagEnd::Heading(level)) => {
                heading_open = None;
                if match_level.is_none() && heading_text.trim() == target {
                    match_level = Some(level);
                    start_byte = Some(range.end);
                } else if let Some(matched) = match_level {
                    if heading_level_rank(level) <= heading_level_rank(matched) {
                        end_byte = Some(range.start);
                        break;
                    }
                }
            }
            Event::Text(text) | Event::Code(text) => {
                if heading_open.is_some() {
                    heading_text.push_str(&text);
                }
            }
            _ => {}
        }
    }

    match (start_byte, end_byte) {
        (Some(s), Some(e)) if e > s => md[s..e].trim().to_string(),
        (Some(s), None) => md[s..].trim().to_string(),
        _ => String::new(),
    }
}

fn heading_level_rank(level: HeadingLevel) -> u8 {
    match level {
        HeadingLevel::H1 => 1,
        HeadingLevel::H2 => 2,
        HeadingLevel::H3 => 3,
        HeadingLevel::H4 => 4,
        HeadingLevel::H5 => 5,
        HeadingLevel::H6 => 6,
    }
}

/// Extract every fenced code-block body. When `lang` is `Some`, only blocks
/// whose info string starts with that language tag (case-sensitive) are
/// returned; when `lang` is `None`, every fenced block is returned.
pub fn extract_code_blocks(md: &str, lang: Option<&str>) -> Vec<String> {
    let parser = Parser::new_ext(md, options());

    let mut blocks: Vec<String> = Vec::new();
    let mut buf = String::new();
    let mut capturing = false;

    for event in parser {
        match event {
            Event::Start(Tag::CodeBlock(kind)) => {
                let info = match kind {
                    CodeBlockKind::Fenced(s) => s.into_string(),
                    CodeBlockKind::Indented => String::new(),
                };
                let info_lang = info.split_whitespace().next().unwrap_or("");
                capturing = match lang {
                    Some(target) => info_lang == target,
                    None => true,
                };
                buf.clear();
            }
            Event::End(TagEnd::CodeBlock) => {
                if capturing {
                    blocks.push(std::mem::take(&mut buf));
                }
                capturing = false;
                buf.clear();
            }
            Event::Text(text) => {
                if capturing {
                    buf.push_str(&text);
                }
            }
            _ => {}
        }
    }

    blocks
}
